### PR TITLE
transformations: Implement unrolled stencil conversion

### PR DESCRIPTION
**Pull request by PapyChacal**
Add support to express and lower unrolled stencils.
It has minimal impact on the general lowering, which leads to a crazy amount of changes in the many filechecks of this lowering.
FWIW, my other PRs are intended to move complicated - and thus tested - logic out of this lowering, so it should alleviate that in the near future.

### CODE REVIEW

File: tests/filecheck/dialects/stencil/hdiff_gpu.mlir [hunk 1]
```diff
@@ -1,73 +1,83 @@
...
+// CHECK-NEXT:        %44 = arith.constant 1 : index
+// CHECK-NEXT:        %45 = arith.addi %20, %44 : index
+// CHECK-NEXT:        %46 = arith.constant 1 : index
+// CHECK-NEXT:        %47 = arith.addi %21, %46 : index
+// CHECK-NEXT:        memref.store %39, %6[%47, %45, %19] : memref<64x64x64xf64, strided<[5184, 72, 1], offset: 21028>>
+// CHECK-NEXT:        scf.yield
+// CHECK-NEXT:      }) : (index, index, index, index, index, index, index, index, index) -> ()
+// CHECK-NEXT:      func.return
+// CHECK-NEXT:    }
+// CHECK-NEXT:  }
```
Comment (superlopuh) [inline tests/filecheck/dialects/stencil/hdiff_gpu.mlir:83]:
```suggestion
// CHECK-NEXT:  }

```

File: xdsl/dialects/stencil.py [hunk 2]
```diff
@@ -747,21 +756,24 @@ def get(res: Sequence[SSAValue | Operation]):
...
-                "stencil.return expected operand types to match the parent "
-                "stencil.apply result element types."
-            )
+        for i, res_type in enumerate(res_types):
+            for type in types[unroll_factor * i : unroll_factor * (i + 1)]:
+                if type != res_type:
+                    raise VerifyException(
+                        "stencil.return expected operand types to match the parent "
+                        "stencil.apply result element types."
+                    )
```
Comment (superlopuh) [inline xdsl/dialects/stencil.py:776]: might be worth mentioning `i` here? or would that be confusing client-side?

Comment (PapyChacal) [reply xdsl/dialects/stencil.py]: Updated, WDYT?

**Review by georgebisbas — CHANGES_REQUESTED** [review]

File: xdsl/dialects/stencil.py [hunk 3]
```diff
@@ -747,21 +756,26 @@ def get(res: Sequence[SSAValue | Operation]):
...
-        if types != res_types:
-            raise VerifyException(
-                "stencil.return expected operand types to match the parent "
-                "stencil.apply result element types."
-            )
+        for i, res_type in enumerate(res_types):
+            for j in range(unroll_factor * i, unroll_factor * (i + 1)):
+                op_type = types[j]
+                if op_type != res_type:
+                    raise VerifyException(
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/dialects/stencil.py:776]: Added a comment, WDYT?

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/stencil.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/stencil.py]: Much clearer, thank you

**Pull request merged**